src/util.ts

```diff
@@ -92,19 +92,33 @@ export const fromAlphabet = (
     if (isSafeValue) {
       return value
     }
-    if (typeof BigInt === 'function') {
-      return BigInt(carry) * BigInt(alphabetChars.length) + BigInt(index)
-    }
-    // we do not have support for BigInt:
-    throw new Error(
+
+    throwIfBigIntNotAvailable(
       `Unable to decode the provided string, due to lack of support for BigInt numbers in the current environment`,
     )
+
+    return BigInt(carry) * BigInt(alphabetChars.length) + BigInt(index)
   }, 0)
 
 const safeToParseNumberRegExp = /^\+?\d+$/
 
-export const safeParseInt10 = (str: string) =>
-  safeToParseNumberRegExp.test(str) ? Number.parseInt(str, 10) : Number.NaN
+export const safeParseInt10 = (str: string) => {
+  if (!safeToParseNumberRegExp.test(str)) {
+    return Number.NaN
+  }
+
+  const int10 = Number.parseInt(str, 10)
+
+  if (Number.isSafeInteger(int10)) {
+    return int10
+  }
+
+  throwIfBigIntNotAvailable(
+    'Unable to encode the provided BigInt string without loss of information due to lack of support for BigInt type in the current environment',
+  )
+
+  return BigInt(int10)
```

Comment:
You can't cast a number to a BigInt safely, since the non-safe integer cast was already was lossy:

> BigInt(Number.parseInt('90071992547409910123456789', 10)) !== 90071992547409910123456789n

Re-parse the original integer as BigInt instead:

```suggestion
  return BigInt(str)
```

Reply:
Got it, fixed!